build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/JarApiComparisonTask.java

```diff
@@ -0,0 +1,195 @@
+                command.add(modulePath);
+            }
+            if (classpath != null) {
+                command.add("--class-path");
+                command.add(classpath);
+            }
+            command.add(location);
+            pb.command(command.toArray(new String[] {}));
+            Process p;
+            try {
+                p = pb.start();
+                p.onExit().get();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+
+            InputStream streamToRead = p.exitValue() == 0 ? p.getInputStream() : p.getErrorStream();
+
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(streamToRead))) {
+                return br.lines().toList();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        /**
+         * Given the output of the javap command, that is, the disassembled class file,
+         * return a set of signatures for all public classes, methods, and fields.
+         */
+        public static Set<String> signaturesSet(List<String> javapOutput) {
+            return javapOutput.stream().filter(s -> s.matches("^\\s*public.*")).collect(Collectors.toSet());
+        }
+
+        /**
+         * Given a disassembled module-info.class, return all unqualified exports.
+         */
+        public static Set<String> moduleInfoSignaturesSet(List<String> javapOutput) {
+            return javapOutput.stream()
+                .filter(s -> s.matches("^\\s*exports.*"))
+                .filter(s -> s.matches(".* to$") == false)
+                .collect(Collectors.toSet());
+        }
+
+        /**
+         * Iterate over classes and gather signatures.
+         */
+        public Map<String, Set<String>> jarSignature() throws IOException {
+            return this.classNames().stream().collect(Collectors.toMap(s -> s, s -> {
+                List<String> disassembled = disassembleFromJar(s, null);
+                if ("module-info.class".equals(s)) {
+                    return moduleInfoSignaturesSet(disassembled);
+                }
+                return signaturesSet(disassembled);
+            }));
+        }
+
+        /**
+         * Comparison: The signatures are maps of class names to public class, field, or method
+         * declarations.
+         * </p>
+         * First, we check that the new jar signature contains all the same classes
+         * as the old jar signature. If not, we return an error.
+         * </p>
+         * Second, we iterate over the signature for each class. If a signature from the old
+         * jar is absent in the new jar, we add it to our list of errors.
+         * </p>
+         * Note that it is fine for the new jar to have additional elements, as this
+         * is backwards compatible.
+         */
+        public static void compareSignatures(Map<String, Set<String>> oldSignature, Map<String, Set<String>> newSignature) {
```

Comment:
There's a few edge cases that naive comparison of `javap` output doesn't take into consideration.

  1. We don't (so far as I can tell) make any consideration as the the visibility of the _class_ to which these members belong is. Perhaps for the most part these projects are all public API and don't have private members but that might not always stricly be the case. Specifically, we don't want to worry about deletions to non-public classes, such as package private classes or private inner classes.
  2. Similarly, we don't ensure that the class we're comparing is exported from the module.
  3. There are categories of changes that would result in error via the simple `javap` string output comparison, but are not necessarily "breaking" changes. This could be something like like changing a method argument type to a superclass or the existing type, or a return type to a subclass of the existing type.
  4. Similarly to 3 any refactor that _moved_ members to another type, which the existing type then extended from would break here to. For example. Class `A` declares a method `foo()`. I  then refactor to move `foo()` to class `B`, which `A` extends. This is not an API breaking change, but it this implementation would flag it as one.

Reply:
I think 4. is a breaking change. I as a plugin developer could still be using A (not the new subclass B) and I'd expect that foo() exists in A. My code would.

I also think 3. is breaking. If we downgrade or upgrade a return type, even if it follows the same contract, a code could be doing an instanceof or a cast, which is now no longer valid.